iup/__init__.py

```diff
@@ -0,0 +1,739 @@
+        self.assert_columns_type(["estimate"], pl.Float64)
+
+    def assert_columns_found(self, columns: List[str]):
+        """
+        Verify that expected columns are found.
+
+        Parameters
+        columns: List[str]
+            names of expected columns
+        """
+        for col in columns:
+            assert col in self.columns, f"Column {col} is expected but not found."
+
+    def assert_columns_type(self, columns: List[str], dtype):
+        """
+        Verify that columns have the expected type.
+
+        Parameters
+        columns: List[str]
+            names of columns for which to check type
+        dtype:
+            data type for each listed column
+        """
+        for col in columns:
+            assert (
+                self[col].dtype == dtype
+            ), f"Column {col} should be {dtype} but is {self[col].dtype}"
+
+    def with_columns(self, *args, **kwargs):
+        """
+        Copy of polars with_columns that returns the same subclass as it's given.
+        """
+        orig_class = type(self)
+        result = super().with_columns(*args, **kwargs)
+        return orig_class(result)
+
+    @staticmethod
+    def date_to_season(date_col: pl.Expr) -> pl.Expr:
+        """
+        Extract season column from a date column, as polars expressions.
+
+        Parameters
+        date_col: pl.Expr
+            column of dates
+
+        Returns
+        pl.Expr
+            column of the season for each date
+
+        Details
+        Assume overwinter seasons, e.g. 2023-10-07 and 2024-04-18 are both in "2023/24"
+        """
+        year1 = (
+            date_col.dt.year() + pl.when(date_col.dt.month() < 7).then(-1).otherwise(0)
+        ).cast(pl.Utf8)
+        year2 = (
+            date_col.dt.year() + pl.when(date_col.dt.month() < 7).then(0).otherwise(1)
+        ).cast(pl.Utf8)
+        season = pl.concat_str([year1, year2], separator="/")
+
+        return season
+
+    @staticmethod
+    def date_to_interval(date_col: pl.Expr) -> pl.Expr:
+        """
+        Extract a time interval column from a date column, as polars expressions.
+
+        Parameters
+        date_col: pl.Expr
+            column of dates
+
+        Returns
+        pl.Expr
+            column of the number of days between each date and the previous
+
+        Details
+        Date column should be chronologically sorted in advance.
+        Time difference is always in days.
+        """
+        interval = date_col.diff().dt.total_days().cast(pl.Float64)
+
+        return interval
+
+    @staticmethod
+    def date_to_elapsed(date_col: pl.Expr) -> pl.Expr:
+        """
+        Extract a time elapsed column from a date column, as polars expressions.
+
+        Parameters
+        date_col: pl.Expr
+            column of dates
+
+        Returns
+        pl.Expr
+            column of the number of days elapsed since the first date
+
+        Details
+        Date column should be chronologically sorted in advance.
```

Comment:
this is something that you could put an assert statement in for

Reply:
Issue #43